sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggUtils.scala

```diff
@@ -135,19 +135,25 @@ object AggUtils {
     }
     val distinctAttributes = namedDistinctExpressions.map(_.toAttribute)
     val groupingAttributes = groupingExpressions.map(_.toAttribute)
+    val filterWithDistinctAttributes = functionsWithDistinct.flatMap(_.filterAttributes.toSeq)
 
     // 1. Create an Aggregate Operator for partial aggregations.
     val partialAggregate: SparkPlan = {
       val aggregateExpressions = functionsWithoutDistinct.map(_.copy(mode = Partial))
       val aggregateAttributes = aggregateExpressions.map(_.resultAttribute)
       // We will group by the original grouping expression, plus an additional expression for the
-      // DISTINCT column. For example, for AVG(DISTINCT value) GROUP BY key, the grouping
-      // expressions will be [key, value].
+      // DISTINCT column and the referred attributes in the FILTER clause associated with each
+      // aggregate function. For example:
+      // 1.for the AVG (DISTINCT value) GROUP BY key, the grouping expression will be [key, value];
+      // 2.for AVG (DISTINCT value) Filter (WHERE value2> 20) GROUP BY key, the grouping expression
+      // will be [key, value, value2].
```

Comment:
I don't get it. The `value` needs to be distinct, so we add it to the grouping expressions to deduplicate. Why does `value2` need to be distinct?

Reply:
For more context, please refer to https://github.com/apache/spark/pull/10228

The first aggregate is to get distinct values, I don't think we should group by the filter column, we just need to evaluate the filter.

Reply:
If we not add value2 to the grouping expressions, the aggregate output will not contains the attribute of value2.

Reply:
But why do we need to output value2? The filter should be done only at the first aggregate.

Reply:
Because the filter is related to aggregate function. The first aggregate and second aggregate not apply the aggregate function, so we need to output value2. The thrid aggregate will apply the aggregate function with input value2.

Reply:
For a query like SELECT COUNT(DISTINCT a) FILTER (WHERE c > 0), SUM(b) FILTER (WHERE d = 0) FROM table will be
```
AGG-4 (count distinct)
  Shuffle to a single reducer
    Partial-AGG-3 (count distinct, no grouping, apply function COUNT on a with c > 0)
      Partial-AGG-2 (grouping on a and c)
        Shuffle by a
          Partial-AGG-1 (grouping on a and c, apply function SUM on b with d = 0)
```

Reply:
Outputting value2 doesn't mean we have to group by value2. We can update the `resultExpressions` to include value2.

Reply:
Oh, I will try to this. Thanks wenchen.

Reply:
I remove value2 from the grouping expressions and find PartialMerge-AGG-3 cannot obtain the input value2. This will lead an error:
```
Caused by: java.lang.RuntimeException: Couldn't find sex#8 in [cast(id#6 as bigint)#13L]
        at scala.sys.package$.error(package.scala:27)
        at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:85)
        at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:79)
        at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
```

Reply:
It seems could not pass value2 to PartialMerge-AGG-3 if remove value2 from the grouping expressions of Partial-AGG-1 or PartialMerge-AGG-2

Reply:
I read the source code, seems the workflow is:
1. local aggregate. **input**: records. **group by**: keys and the distinct column. **output**: agg buffers
2. shuffle aggregate. **input**: agg buffers. **group by**: keys and the distinct column. **output**: agg buffers
3. local aggregate. **input**: agg buffers. **group by**: keys. **output**: agg buffers
4. shuffle aggregate. **input**: agg buffers. **group by**: keys. **output**: agg buffers

So I think we only need to evaluate the filter in the first aggregate.

Reply:
Apply filter must go with applying aggregate function . 
The first aggregate only apply the aggregate function that not contains DISTINCT.
The third aggregate will apply the aggregate function that contains DISTINCT .

Reply:
@cloud-fan I modified `RewriteDistinctAggregates`.
If there exists one distinct aggregate function, will still uses `planAggregateWithOneDistinct`.
If there exists more than one distinct aggregate function, will still uses `RewriteDistinctAggregates`.

Reply:
Ah, I see... It seems we cannot simply filter out aggregate input rows in the first (aggregate) phase because each aggregate function can have different filter conditions and their actual aggregate processing can happen in different phases. I think we need to study more about this impl. and add exhaustive tests for distinct cases.

So, how about dropping support `distinct + filter aggregate` in this pr and moving this discussion into a next PR to support `distinct + filter aggregate`? WDYT, @cloud-fan ? We should include the support in this pr?

Reply:
@maropu I have modified `RewriteDistinctAggregates`.
If query exists more than one distinct aggregate function, then the filter will apply to second aggregate.
If query exists only one distinct aggregate function, then the filter will apply to third aggregate.